### ISSUE TITLE
Need to be more explicit about ORKY_URI

### ISSUE THREAD

**Issue by blag** [OfficeDev/Orky#21]
# Background and Question #

In step 4 of the [Instructions](https://github.com/OfficeDev/Orky/tree/master/Hubot-Orky#instructions) the `ORKY_URI` environment variable is set to `<path to your Orky instance>`.

But the path to the Orky instance isn't explicitly discussed anywhere.

The only URI that is mentioned in any of the instructions (that I have found) is from the [instructions for setting up Orky on Azure](https://github.com/OfficeDev/Orky/tree/master/Orky#create-a-new-azure-web-app): `https://orkydemo.azurewebsites.net/api/messages` endpoint.

Is that what I should use for the path to the Orky instance? It would be very helpful if the README explains where the Orky instance URI comes from.

## Things I have tried ##

I have tried the following values (setup with my Orky instance) with the following hubot logs/results:

#### `ORKY_URI=https://<orky-instance>.azurewebsites.net/api/messages` ####

```
[Thu Mar 14 2019 17:50:44 GMT+0000 (Coordinated Universal Time)] DEBUG Loading adapter orky
[Thu Mar 14 2019 17:50:45 GMT+0000 (Coordinated Universal Time)] INFO Created instance of Orky Adapter with config: {
  "OrkyUri": "https://<orky-instance>.azurewebsites.net/api/messages",
  "BotId": "<orky-bot-id>",
  "BotSecret": "<redacted>"
}
[Thu Mar 14 2019 17:50:45 GMT+0000 (Coordinated Universal Time)] INFO Run
[Thu Mar 14 2019 17:50:46 GMT+0000 (Coordinated Universal Time)] ERROR Socket.io received an error. error="Invalid namespace"
[Thu Mar 14 2019 17:51:11 GMT+0000 (Coordinated Universal Time)] DEBUG Sent a ping to the server
[Thu Mar 14 2019 17:51:11 GMT+0000 (Coordinated Universal Time)] DEBUG Received a pong from the server. Latency=43ms
```

The ping/pong continuously repeats.

#### `ORKY_URI=https://<orky-instance>.azurewebsites.net` ####

```
[Thu Mar 14 2019 18:07:30 GMT+0000 (Coordinated Universal Time)] DEBUG Loading adapter orky
[Thu Mar 14 2019 18:07:31 GMT+0000 (Coordinated Universal Time)] INFO Created instance of Orky Adapter with config: {
  "OrkyUri": "https://<orky-instance>.azurewebsites.net",
  "BotId": "<orky-bot-id>",
  "BotSecret": "<redacted>"
}
[Thu Mar 14 2019 18:07:31 GMT+0000 (Coordinated Universal Time)] INFO Run
[Thu Mar 14 2019 18:07:32 GMT+0000 (Coordinated Universal Time)] INFO Connected to Orky server at https://<orky-instance>.azurewebsites.net
[Thu Mar 14 2019 18:07:32 GMT+0000 (Coordinated Universal Time)] ERROR Registration details are incorrect.
```

Logs end.

#### `ORKY_URI=https://<orky-instance>.azurewebsites.net/api` ####

```
[Thu Mar 14 2019 18:08:05 GMT+0000 (Coordinated Universal Time)] DEBUG Loading adapter orky
[Thu Mar 14 2019 18:08:06 GMT+0000 (Coordinated Universal Time)] INFO Created instance of Orky Adapter with config: {
  "OrkyUri": "https://<orky-instance>.azurewebsites.net/api",
  "BotId": "<orky-bot-id>",
  "BotSecret": "<redacted>"
}
[Thu Mar 14 2019 18:08:06 GMT+0000 (Coordinated Universal Time)] INFO Run
[Thu Mar 14 2019 18:08:07 GMT+0000 (Coordinated Universal Time)] ERROR Socket.io received an error. error="Invalid namespace"
[Thu Mar 14 2019 18:08:31 GMT+0000 (Coordinated Universal Time)] DEBUG Sent a ping to the server
[Thu Mar 14 2019 18:08:32 GMT+0000 (Coordinated Universal Time)] DEBUG Received a pong from the server. Latency=181ms
```

The ping/pong continuously repeats.

## Troubleshooting information ##

I have followed all instructions for setting up an Orky instance using a Web App on Azure, and I am running hubot-orky locally on my laptop.

This is the (redacted) hubot settings:

```
export HUBOT_ADAPTER=orky
export ORKY_URI=https://<orky-instance>.azurewebsites.net/api
# ^^ Also tried
# https://<orky-instance>.azurewebsites.net/api/messages
# and
# https://<orky-instance>.azurewebsites.net
export BOT_ID=<orky-bot-id>
export BOT_SECRET=<redacted>
```

Sending messages in the BotFramework bot test page all successfully send, and I can see the requests count increasing in the Azure web app metrics page, but nothing appears in the local hubot logs.

I have also sideloaded a modified version of Orky into my MS Teams instance according to the instructions [here](https://github.com/OfficeDev/Orky/tree/master/Orky#create-an-application-package-and-sideload-it). Sending messages to the bot via MS Teams also succeeds, and also increments the requests metric in the Azure web app metrics page.

Due to the `ERROR` from the hubot logs, and the lack of logged messages, I believe my local hubot instance is not successfully connected to the Orky instance on Azure.

**Comment by blag** [comment]
I have seen the same behavior with:
* `ORKY_URI=https://<orky-instance>.azurewebsites.net/socket.io`
  ```
  [Thu Mar 14 2019 20:42:33 GMT+0000 (Coordinated Universal Time)] DEBUG Loading adapter orky
  [Thu Mar 14 2019 20:42:34 GMT+0000 (Coordinated Universal Time)] INFO Created instance of Orky Adapter with config: {
    "OrkyUri": "https://<orky-instance>.azurewebsites.net/hubot/socket.io",
    "BotId": "<orky-bot-id>",
    "BotSecret": "<redacted>"
  }
  [Thu Mar 14 2019 20:42:34 GMT+0000 (Coordinated Universal Time)] INFO Run
  [Thu Mar 14 2019 20:42:35 GMT+0000 (Coordinated Universal Time)] ERROR Socket.io received an error. error="Invalid namespace"
  [Thu Mar 14 2019 20:42:59 GMT+0000 (Coordinated Universal Time)] DEBUG Sent a ping to the server
  [Thu Mar 14 2019 20:42:59 GMT+0000 (Coordinated Universal Time)] DEBUG Received a pong from the server. Latency=45ms
  ```
* `ORKY_URI=https://<orky-instance>.azurewebsites.net/hubot/socket.io`
  ```
  [Thu Mar 14 2019 20:49:07 GMT+0000 (Coordinated Universal Time)] DEBUG Loading adapter orky
  [Thu Mar 14 2019 20:49:08 GMT+0000 (Coordinated Universal Time)] INFO Created instance of Orky Adapter with config: {
    "OrkyUri": "https://<orky-instance>.azurewebsites.net/hubot/socket.io",
    "BotId": "<orky-bot-id>",
    "BotSecret": "<redacted>"
  }
  [Thu Mar 14 2019 20:49:08 GMT+0000 (Coordinated Universal Time)] INFO Run
  [Thu Mar 14 2019 20:49:09 GMT+0000 (Coordinated Universal Time)] ERROR Socket.io received an error. error="Invalid namespace"
  [Thu Mar 14 2019 20:49:33 GMT+0000 (Coordinated Universal Time)] DEBUG Sent a ping to the server
  [Thu Mar 14 2019 20:49:33 GMT+0000 (Coordinated Universal Time)] DEBUG Received a pong from the server. Latency=44ms
  ```

**Comment by MattSFT** [comment]
It should be the first one you have with https://<orky-instance>.azurewebsites.net/socket.io
You also need to turn on web sockets in azure:
![image](https://user-images.githubusercontent.com/29067893/54392146-18e75b00-4664-11e9-8390-cbda16aaa83c.png)

**Comment by blag** [comment]
Thank you for confirming.

As per the directions, I do have Web Sockets turned on for the Azure Web App, as well as Always On.

**Comment by blag** [comment]
I set `LOG_LEVEL` to `debug` for the Azure Web App and restarted it.

From the logs:

```
2019-03-14T21:32:52.329Z [debug] Connection (<redacted>) did not attempt to register.
2019-03-14T21:32:52.329Z Connection is not authorized.
2019-03-14T21:32:52.329Z Error: Connection is not authorized.
    at unauthorized (D:\home\site\wwwroot\build\Services\BotConnectionManager.js:96:31)
    at Timeout.setTimeout [as _onTimeout] (D:\home\site\wwwroot\build\Services\BotConnectionManager.js:123:28)
    at ontimeout (timers.js:482:11)
    at tryOnTimeout (timers.js:317:5)
    at Timer.listOnTimeout (timers.js:277:5)
```

**Comment by blag** [comment]
After setting
`export ORKY_URI=https://<orky-instance>.azurewebsites.net`

My local hubot instance can at least connect to the Orky instance:

```
[Thu Mar 14 2019 23:32:45 GMT+0000 (Coordinated Universal Time)] DEBUG Loading adapter orky
[Thu Mar 14 2019 23:32:46 GMT+0000 (Coordinated Universal Time)] INFO Created instance of Orky Adapter with config: {
  "OrkyUri": "https://<orky-instance>.azurewebsites.net",
  "BotId": "<orky-bot-id>",
  "BotSecret": "<redacted>"
}
[Thu Mar 14 2019 23:32:46 GMT+0000 (Coordinated Universal Time)] INFO Run
[Thu Mar 14 2019 23:32:47 GMT+0000 (Coordinated Universal Time)] INFO Connected to Orky server at https://<orky-instance>.azurewebsites.net
[Thu Mar 14 2019 23:32:47 GMT+0000 (Coordinated Universal Time)] ERROR Registration details are incorrect.
```

I believe I have tracked down what the correct value of `ORKY_URI` should be, and more importantly, what it _cannot_ be.

In `Orky.coffee` from hubot-orky, it [initializes a socket](https://github.com/OfficeDev/Orky/blob/master/Hubot-Orky/src/Orky.coffee#L31) with just the URI grabbed from `ORKY_URI`:

```coffee
@config =
  OrkyUri: process.env.ORKY_URI
  BotId: process.env.BOT_ID
  BotSecret: process.env.BOT_SECRET
```

and passes the `OrkyUri` to `SocketIO()`:

```
@client = SocketIO(@config.OrkyUri)
```

which means the [socketio-client's `lookup` function](https://github.com/socketio/socket.io-client/blob/master/lib/index.js#L36) doesn't receive an `opts` parameter:

```js
function lookup (uri, opts) {
  if (typeof uri === 'object') {  // typeof uri will be string
    // not run
  }

  opts = opts || {};  // opts is undefined at this line, so it gets assigned to {}

  var parsed = url(uri);
  var source = parsed.source;  // literally just uri (in its entirety) again
```

and later on the `source`, which is the validated `OrkyUri` URI, gets passed to `Manager()`:

```
  ... = Manager(source, opts);  // source is the parsed URL from the uri, and opts is {}
```

so the [`Manager function`](https://github.com/socketio/socket.io-client/blob/8372591652260f88987c467b94573a144ee4db63/lib/manager.js#L36) receives the `source` and an empty object (`{}`) for `opts`:

```js
function Manager (uri, opts) {
  if (!(this instanceof Manager)) return new Manager(uri, opts);
  if (uri && ('object' === typeof uri)) {  // typeof uri is still string
    // not run
  }
  opts = opts || {};  // opts was originally {}, so it stays {} here

  opts.path = opts.path || '/socket.io';  // opts.path gets set to '/socket.io'

  ...

  this.opts = opts;
```

All this means that the correct value for the `ORKY_URI` variable is the **URL** (yoo-arr-**ell**) to the Orky instance. The `/socket.io` path will unconditionally be appended to the value, so no path should be specified.

If users followed the [linked instructions](https://github.com/OfficeDev/Orky/tree/master/Hubot-Orky#what-is-orky) (see the last link in the second sentence) to setup the Orky instance as an Azure Web App, then the proper setting for `ORKY_URI` would simply be:

`export ORKY_URI=https://<orky-instance>.azurewebsites.net`

**Comment by MattSFT** [comment]
Thanks for investigating this! Would this change have helped make this easier for you? https://github.com/OfficeDev/Orky/pull/22

Feel free to propose a different file change.

**Comment by blag** [comment]
Yes, I think it would have saved me some time.

I'm writing a PR, to fix this issue and others.

**Comment by MattSFT** [comment]
ok cool. I'll abandon mine and wait for yours to come through!

**Comment by blag** [comment]
Fixed in #23. I was originally planning on using Orky, but I couldn't get it working acceptably, so I ended up using BotFramework directly.

I would still be interested in Orky, but it needs to mature a bit more before I can use it. Thank you for all of your work on this project and your help on this issue!

Closing.
